cpp/unreal_plugins/UrdfRobot/Source/UrdfRobot/UrdfRobotComponent.cpp

```diff
@@ -0,0 +1,313 @@
+#include "CoreUtils/InputActionComponent.h"
+#include "CoreUtils/Log.h"
+#include "CoreUtils/Std.h"
+#include "CoreUtils/Unreal.h"
+#include "UrdfRobot/UrdfJointComponent.h"
+#include "UrdfRobot/UrdfLinkComponent.h"
+#include "UrdfRobot/UrdfParser.h"
+
+// useful for debugging horizontal_pendulum.urdf
+const std::map<std::string, std::map<std::string, std::vector<double>>> DEFAULT_INPUT_ACTIONS = {
+    {"One",   {{"joint_0.add_torque_in_radians",          { 1000000.0, 0.0, 0.0}}}},
+    {"Two",   {{"joint_0.add_torque_in_radians",          {-1000000.0, 0.0, 0.0}}}},
+    {"Three", {{"joint_1.add_to_angular_velocity_target", { 0.1,       0.0, 0.0}}}},
+    {"Four",  {{"joint_1.add_to_angular_velocity_target", {-0.1,       0.0, 0.0}}}}
+};
+
+UUrdfRobotComponent::UUrdfRobotComponent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    // In the current setup, the UrdfRobotComponent and the UrdfBotPawn (UrdfRobotComponent is the RootComponent of this Pawn) 
+    // do not move along with the movement of child LinkComponents. However, we want the UrdfBotPawn's pose to update/follow 
+    // the pose of child LinkComponents. Hence, we enable this component to tick and in every tick we update it's pose to 
+    // follow the root link's pose.
+    PrimaryComponentTick.bCanEverTick = true;
+    PrimaryComponentTick.TickGroup = ETickingGroup::TG_PostPhysics;
+
+    // UInputActionComponent
+    input_action_component_ = CreateDefaultSubobject<UInputActionComponent>(Unreal::toFName("input_action_component"));
+    SP_ASSERT(input_action_component_);
+    // Need to explicitly set this up so that the component hierarchy is well-defined.
+    input_action_component_->SetupAttachment(this);
+}
+
+UUrdfRobotComponent::~UUrdfRobotComponent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    // Objects created with CreateDefaultSubobject, DuplicateObject, LoadObject, NewObject don't need to be cleaned up explicitly.
+
+    LinkComponents.Empty();
+    JointComponents.Empty();
+
+    link_components_.clear();
+    joint_components_.clear();
+
+    SP_ASSERT(input_action_component_);
+    input_action_component_ = nullptr;
+}
+
+void UUrdfRobotComponent::BeginPlay()
+{
+    USceneComponent::BeginPlay();
+
+    // Get player input actions from the config system if it is initialized, otherwise use hard-coded keyboard actions, which
+    // can be useful for debugging.
+    std::map<std::string, std::map<std::string, std::vector<double>>> input_actions;
+    if (Config::s_initialized_) {
+        input_actions = Config::get<std::map<std::string, std::map<std::string, std::vector<double>>>>("URDF_ROBOT.URDF_ROBOT_COMPONENT.INPUT_ACTIONS");
+    } else {
+        input_actions = DEFAULT_INPUT_ACTIONS;
+    }
+
+    input_action_component_->bindInputActions(input_actions);
+    input_action_component_->apply_input_action_func_ = [this, input_actions](const std::string& key) -> void {
+        if (EnableKeyboardControl) {
+            // only assert if we're not in the editor
```

Comment:
```
            // use private applyAction method because it will print warnings rather than assert,
            // and this is desirable when we're working in the editor
            applyAction(input_actions.at(key));
```

Reply:
Done